src/video/windows/SDL_windowsevents.c

```diff
@@ -1533,37 +1533,124 @@ WIN_SendWakeupEvent(_THIS, SDL_Window *window)
+            if (msg.message == WM_MOUSEMOVE) {
+                if (GetMouseMessageSource() == SDL_MOUSE_EVENT_SOURCE_MOUSE) {
+                    /* We should only care about the last real mouse movement. */
+                    mousemove = msg;
+                    mousemove_new = SDL_TRUE;
+                    /* Remove everything that was added to the queue before we started this loop,
+                       treat anything added afterwards as auto-generated. */
+                    if (SDL_TICKS_PASSED(msg.time, start_ticks)) {
+                        /* Very likely an autogenerated mouse move, indicating the end of the normal queue.
+                           Anything sent or posted by an overlay should've been handled by now.
+                           Let's make sure to hit N consecutive mouse moves before dropping out too soon though... */
+                        ++mousemoves;
+                        if (mousemoves >= MAX_MOUSEMOVES_PER_PUMP) {
+                            break;
+                        }
+                        continue;
+                    } else {
+                        /* Let's check if there are any newer mouse moves which are worth handling instead. */
+                        continue;
+                    }
+                } else {
+                    /* Non-mouse mouse movement? Let's handle it normally. */
+                    mousemoves = 0;
+                }
             }
 
-            /* Always translate the message in case it's a non-SDL window (e.g. with Qt integration) */
-            TranslateMessage(&msg);
-            DispatchMessage(&msg);
+            /* Handle all other messages normally and reset the mousemove-only pump counter. */
+            data->mousemove_only_pumps = 0;
+            if (WIN_TranslateDispatchMessage(&msg, start_ticks, &new_messages)) {
+                break;
+            }
+        }
 
-            /* Make sure we don't busy loop here forever if there are lots of events coming in */
-            if (SDL_TICKS_PASSED(msg.time, start_ticks)) {
-                /* We might get a few new messages generated by the Steam overlay or other application hooks
-                   In this case those messages will be processed before any pending input, so we want to continue after those messages.
-                   (thanks to Peter Deayton for his investigation here)
-                 */
-                const int MAX_NEW_MESSAGES = 3;
-                ++new_messages;
-                if (new_messages > MAX_NEW_MESSAGES) {
-                    break;
-                }
+        /* Process the last WM_MOUSEMOVE, but only N consecutive mousemove-only pumps. */
+        if (mousemove_new) {
+            ++data->mousemove_only_pumps;
+            if (data->mousemove_only_pumps < MAX_MOUSEMOVE_PUMPS) {
+                WIN_TranslateDispatchMessage(&mousemove, start_ticks, &new_messages);
```

Comment:
Dispatching messages out of order seems like a very bad idea. Functions which retrieve extra message data (`GetMessage[Pos/Time/ExtraInfo]`) are based on the `the last message retrieved by the GetMessage function`.

Reply:
Hm, I was under the naive assumption that mouse moves will end up being the last messages in the queue among other auto-generated messages like WM_PAINT, but now I'm feeling even less safe with that approach (especially given how I forgot about `Pos` and `ExtraInfo`). I'll try switching to a `WIN_WindowProc`-based approach after all 😅 hopefully it'll end up prettier than what I've got on my mind right now though. Thanks :)

Reply:
I've changed my approach so that only SDL's own handling (which results in either `SDL_SendMouseMotion` or `SDL_SetMouseFocus`) ends up being delayed without affecting the execution order of hooks or custom `WndProc`s, while keeping the part in the message loop itself which looks out for "mouse move only pumps".